### DISCUSSION TITLE
Adding Computed Property Destructuring to Vue

### DISCUSSION THREAD

**Discussion by bosens-China** [vuejs/rfcs#729] [RFC Discussions]
## **Background**  
In Vue 3, `computed` properties are commonly used as reactive tools to derive new data from existing state. However, when a `computed` property returns an object, directly destructuring it results in the loss of reactivity for the destructured properties. To maintain reactivity, developers currently have to manually create new `computed` properties for each field, as shown below:  

```javascript
const x = computed(() => ({ a: 1, b: 2 }));
const a = computed(() => x.value.a); // Manually create for each property
const b = computed(() => x.value.b);
```

This approach becomes verbose and cumbersome, especially when dealing with objects with many properties or deeply nested structures. It reduces code readability and development efficiency.  

---

## **Significance and Usage**  
**Significance**:  
1. **Simplified Code**: Supporting direct destructuring of computed properties can eliminate redundant boilerplate code and streamline logic.  
2. **Improved Readability**: Developers can destructure `computed` properties like regular objects, while maintaining reactivity, which enhances code clarity.  
3. **Support for Nested Structures**: Not only for flat objects, but it should also allow reactive destructuring of deeply nested structures, reducing the complexity of manual handling.  

**Usage**:  
With the proposed improvement, computed properties can be destructured directly. Here's an example:  

```javascript
const { a, b } = computed(() => ({ a: 1, b: 2 }));
console.log(a.value); // 1
console.log(b.value); // 2
```

**Support for Nested Destructuring**:  
If a computed property returns a nested object, it should also allow nested destructuring, with all destructured properties remaining reactive:  

```javascript
const { user: { name, age }, isActive } = computed(() => ({
  user: { name: 'Alice', age: 25 },
  isActive: true,
}));

console.log(name.value); // 'Alice'
console.log(age.value);  // 25
console.log(isActive.value); // true
```

**Completeness and Compatibility**:  
- All destructured properties remain fully reactive and compatible with Vue’s reactivity system.  
- This unified approach would work seamlessly for both shallow and deeply nested structures, aligning with developers’ intuitive usage.  

By introducing this feature, Vue’s `computed` properties would become even more developer-friendly, reducing redundancy and maintenance overhead while enhancing productivity and code readability.

**Comment by Demivan** [comment]
This would be impossible to type with TypeScript.

You would need `computed` to return a different object when destructuring.

```ts
const x = computed(() => ({ a: 1, b: 2 }));
// { value: { a: number, b: number } }

const { a, b } = computed(() => ({ a: 1, b: 2 }));
// { a: { value: number }, b: { value: number } }
```

**Reply by bosens-China** [reply]
> This would be impossible to type with TypeScript.
> 
> You would need `computed` to return a different object when destructuring.
> 
> ```ts
> const x = computed(() => ({ a: 1, b: 2 }));
> // { value: { a: number, b: number } }
> 
> const { a, b } = computed(() => ({ a: 1, b: 2 }));
> // { a: { value: number }, b: { value: number } }
> ```

I think it's possible to restrict some of the usage, for example, there are constraints on the definition of props now

**Comment by posva** [comment]
This has a few deal breakers: it's a huge breaking change, it an inconsistent behavior, and can be [implemented in user land](https://play.vuejs.org/#eNqNVFFvm0AM/isWL0CVkYftKaNo3VRNW7V16tBeepVyISahhbvT3dEmQvz3+Y6QoGab9hLA/mx//mynC66USp5bDBZBagpdKQsGbaug5mJzyQJrWJAxUTVKagsdFLJRrcU19FBq2UBIseH7CcDKTweIOWKS+ckaMsFEIYVxYD6DFaEup1HRWCKKYrjMICLYAkIeEpaeqxD6OI5dlnQ+MCZ+9GGxUTW3SF8AqdKYdR3ve+g6V6NP587kgibAYEYNEpmy2iSPRgpSoXPhLHAkqhr1rbIVkWXBArzH+Xhdy5ev3mZ1i7PRbqyuCntmLrZYPP0B/mh2zsaCHxoN6mdkwdFnud6gT8WC65/fcUfvR2cj121N6H8479DIunXUB9jHVqypmwnON/HFT60Sm9xc7ywKM/Z6JNr7XxbQmN2E/ibJie/b5J2PY6J36h7nmlhD6k7W5BevqQbYvUIYQXdYHizf+H5FXZS3+jNai3p2tnmeEwto95iYX1xQ0QvINRemlLoxwGE5RqR5tgSpgRhSJlgOe5UvoRJWAhcgV49YWJDlqYjSUiFJgybxmT8ornkDRra6QHgDV1C2onBCgN1yC5qORgsqO83mO8kPCUbE1RHxskWNgLzYwhPuXUAODVeGxCH6E018LleFYKGBZ6eczzpnAnde0ROb0yWlOdBuoKBLvMNC6nXqNlRsZsRyn2WRm9PQ0OJMb9KMJh/TiOH+hoRypR3Bh8WUWJrf3zxkNBG/CMNV0zK3taWb7nrg5v/i/RABbr0uCUFNdFiQaCAYxwmN9ZqkiiJyL8Z0fpKHLRwK35Pngaq/+hd5lc6jYipLm+oe7mUY0CEN2WiFg/43h1m8VQ==) (Nested support can be achieved with Proxies). It could maybe also be a compile time but would require a different kind of `computed()` for types. If you want to go further I would suggest you to write a version that works and propose it as a PR at https://vue-macros.dev

**Reply by bosens-China** [reply]
> This has a few deal breakers: it's a huge breaking change, it an inconsistent behavior, and can be [implemented in user land](https://play.vuejs.org/#eNqNVFFvm0AM/isWL0CVkYftKaNo3VRNW7V16tBeepVyISahhbvT3dEmQvz3+Y6QoGab9hLA/mx//mynC66USp5bDBZBagpdKQsGbaug5mJzyQJrWJAxUTVKagsdFLJRrcU19FBq2UBIseH7CcDKTweIOWKS+ckaMsFEIYVxYD6DFaEup1HRWCKKYrjMICLYAkIeEpaeqxD6OI5dlnQ+MCZ+9GGxUTW3SF8AqdKYdR3ve+g6V6NP587kgibAYEYNEpmy2iSPRgpSoXPhLHAkqhr1rbIVkWXBArzH+Xhdy5ev3mZ1i7PRbqyuCntmLrZYPP0B/mh2zsaCHxoN6mdkwdFnud6gT8WC65/fcUfvR2cj121N6H8479DIunXUB9jHVqypmwnON/HFT60Sm9xc7ywKM/Z6JNr7XxbQmN2E/ibJie/b5J2PY6J36h7nmlhD6k7W5BevqQbYvUIYQXdYHizf+H5FXZS3+jNai3p2tnmeEwto95iYX1xQ0QvINRemlLoxwGE5RqR5tgSpgRhSJlgOe5UvoRJWAhcgV49YWJDlqYjSUiFJgybxmT8ornkDRra6QHgDV1C2onBCgN1yC5qORgsqO83mO8kPCUbE1RHxskWNgLzYwhPuXUAODVeGxCH6E018LleFYKGBZ6eczzpnAnde0ROb0yWlOdBuoKBLvMNC6nXqNlRsZsRyn2WRm9PQ0OJMb9KMJh/TiOH+hoRypR3Bh8WUWJrf3zxkNBG/CMNV0zK3taWb7nrg5v/i/RABbr0uCUFNdFiQaCAYxwmN9ZqkiiJyL8Z0fpKHLRwK35Pngaq/+hd5lc6jYipLm+oe7mUY0CEN2WiFg/43h1m8VQ==) (Nested support can be achieved with Proxies). It could maybe also be a compile time but would require a different kind of `computed()` for types. If you want to go further I would suggest you to write a version that works and propose it as a PR at https://vue-macros.dev

After reading the release of version 3.5, I personally think that as a compilation macro, it is more suitable and can also reduce incompatibility.
Moreover, it will not cause any additional performance loss
